### PR TITLE
Include cross-signing signatures when syncing remote devices for the first time

### PR DESCRIPTION
**Pull request by erikjohnston**
When fetching remote devices for the first time, we did not correctly include the cross signing keys in the returned results. 

Reviewable commit by commit.

I think replaces #10912

cc #11159

### CODE REVIEW

File: tests/handlers/test_e2e_keys.py [hunk 1]
```diff
@@ -630,3 +632,151 @@ def test_upload_signatures(self):
...
+                    "keys": {
+                        "ed25519:" + remote_self_signing_key: remote_self_signing_key
+                    },
+                }
+            },
+        )
+
+    def test_query_devices_remote_sync(self):
+        """Tests that querying keys for a remote user that we don't share a room
+        with returns the cross signing keys correctly.
```
Comment (DMRobertson) [inline tests/handlers/test_e2e_keys.py:711]: This is the same docstring as above. How does this test differ to the previous one?

File: synapse/handlers/e2e_keys.py [hunk 2]
```diff
@@ -278,19 +278,30 @@ async def _query_devices_for_destination(
...
                 for device in user_devices:
                     user_results[device["device_id"]] = device["keys"]
                 user_ids_updated.append(user_id)
+
+                # Add any cross signing keys to the results.
+                master_key = resync_results.get("master_key")
+                self_signing_key = resync_results.get("self_signing_key")
+
+                if master_key:
+                    cross_signing_keys["master_keys"][user_id] = master_key
```
Comment (DMRobertson) [inline synapse/handlers/e2e_keys.py:301]: Looks like this is an out parameter? Feels a bit icky.

**Review by DMRobertson — COMMENTED** [review]
Would like another pair of eyes please!

File: tests/handlers/test_e2e_keys.py (hunk 1 above)
Comment (erikjohnston) [reply tests/handlers/test_e2e_keys.py:711]: Woops

File: synapse/handlers/e2e_keys.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/e2e_keys.py:301]: Ugh, yes. The alternative is to return a bunch of dicts, but that is then a bit convoluted to handle when using `defer.gatherResults`. I mean its fine, but I didn't really want to rewrite it too much

**Review by anoadragon453 — COMMENTED** [review]
Looks mostly good to me, and the included refactor is welcome. Just some minor things really.

File: tests/handlers/test_e2e_keys.py [hunk 3]
```diff
@@ -630,3 +632,151 @@ def test_upload_signatures(self):
...
+        """
+
+        remote_user_id = "@test:other"
+        local_user_id = "@test:test"
+
+        remote_master_key = "85T7JXPFBAySB/jwby4S3lBPTqY3+Zg53nYuGmu1ggY"
+        remote_self_signing_key = "QeIiFEjluPBtI7WQdG365QKZcFs9kqmHir6RBD0//nQ"
+
+        self.hs.get_federation_client().query_client_keys = mock.Mock(
+            return_value=defer.succeed(
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply tests/handlers/test_e2e_keys.py:648]: Yeah, nothing should care about the difference between a deferred and coroutine there.

File: tests/handlers/test_e2e_keys.py [hunk 4]
```diff
@@ -630,3 +632,151 @@ def test_upload_signatures(self):
...
+        )
+
+        self.assertEqual(query_result["failures"], {})
+        self.assertEqual(
+            query_result["master_keys"],
+            {
+                remote_user_id: {
+                    "user_id": remote_user_id,
+                    "usage": ["master"],
+                    "keys": {"ed25519:" + remote_master_key: remote_master_key},
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply tests/handlers/test_e2e_keys.py:768]: :shrug:

**Review by anoadragon453 — APPROVED** [review]
lgtm.

**Pull request merged**